### PR TITLE
wip: adding in ambr parser for edd import

### PR DESCRIPTION
**Pull request by somtirtharoy**

### CODE REVIEW

File: server/edd/load/parsers/core.py [hunk 1]
```diff
@@ -641,6 +643,104 @@ def _raw_cell_value(self, cell):
         return val
 
 
+class MultiSheetExcelParserMixin:
```
Comment (wmorrell) [inline server/edd/load/parsers/core.py:648]: This is a bad name, given what the class actually does. If you're going to make a mixin class for handling multiple sheets, the class should *only* concern itself with handling multiple sheets. It should delegate responsibility for interpreting the *contents* of those sheets to a different class.

File: Pipfile [hunk 2]
```diff
@@ -0,0 +1,52 @@
+[[source]]
```
Comment (wmorrell) [inline Pipfile]: Delete this file. The `Pipfile` used to install packages lives next to the `Dockerfile` for the core image, under `docker/edd/core/`.

File: Pipfile.lock [hunk 3]
```diff
@@ -0,0 +1,1553 @@
+{
```
Comment (wmorrell) [inline Pipfile.lock]: Delete this file. The `Pipfile.lock` used to lock packages lives next to the `Dockerfile` for the core image, under `docker/edd/core/`.

File: server/edd/load/parsers/ambr.py [hunk 4]
```diff
@@ -0,0 +1,6 @@
+from .core import MultiSheetExcelParserMixin
+from .generic import GenericImportParser
+
+
+class AmbrExcelParser(MultiSheetExcelParserMixin, GenericImportParser):
```
Comment (wmorrell) [inline server/edd/load/parsers/ambr.py:5]: Ambr does not use the Generic spreadsheet layout. Mixing up unlike things this way will only cause bugs and headaches down the line.

File: server/edd/load/parsers/core.py [hunk 5]
```diff
@@ -641,6 +643,104 @@ def _raw_cell_value(self, cell):
...
+
+    loa_name: str
+    parsed_df: pd.DataFrame
+    df: pd.DataFrame
+    units: Dict
+
+    def __init__(self, sheet_name, df):
+        self.loa_name = sheet_name
+        self.df = df
+        self.units = {
```
Comment (wmorrell) [inline server/edd/load/parsers/core.py]: Do not hard-code these. All of this must be configuration in the database. The example file you've been using is *not* complete, there are hundreds of possible columns that could be included in any given import. When you hard-code it instead of looking up values from database configuration, you're forcing a code change whenever someone decides to check a slightly different column when exporting from the instrument.

File: server/edd/load/parsers/core.py [hunk 6]
```diff
@@ -641,6 +643,104 @@ def _raw_cell_value(self, cell):
...
+
+        mtype_name = self.df[self.df.columns[1:2]].columns.values[0]
+        self.df["Line Name"] = self.loa_name
+        self.df.columns.values[0] = "Time"
+        self.df.columns.values[1] = "Value"
+        self.df["Measurement Type"] = mtype_name
+        self.df["Units"] = self.units[mtype_name]
+        # dropping records with NaN values
+        self.df = self.df[self.df["Value"].notna()]
+        self.parsed_df = self.df
```
Comment (wmorrell) [inline server/edd/load/parsers/core.py]: It makes no sense to have two attributes with the same value. You want a boolean flag for whether the dataframe is "parsed" or not, instead of relying on another attribute having a dataframe or `None`.

File: server/edd/load/parsers/core.py [hunk 7]
```diff
@@ -641,6 +643,104 @@ def _raw_cell_value(self, cell):
...
+
+        wb = pd.read_excel(file, sheet_name=None)
+
+        # for each worksheet in the workbook
+        parsed_result = pd.DataFrame()
+
+        for name, sheet in wb.items():
+            # for every two columns in the worksheet
+            # corresponding to each measurement type in the sheet
+            for i in range(0, int(sheet.shape[1]), 2):
```
Comment (wmorrell) [inline server/edd/load/parsers/core.py]: This loop should be the core of an Ambr parser class. First column sample name, mapped to either an existing assay or a created assay linked to a line. Then pairs of columns for values of a specific measurement and the time of that measurement.

File: server/edd/load/parsers/demo.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
+import numbers
```
Comment (wmorrell) [inline server/edd/load/parsers/demo.py]: What is the purpose of this file being stand-alone? Just turn it into normal test code and run it with the normal test facilities. Look up the `-k` flag of `pytest`.

File: server/edd/load/parsers/generic.py [hunk 9]
```diff
@@ -95,3 +95,5 @@ class GenericExcelParser(ExcelParserMixin, GenericImportParser):
 
 class GenericCsvParser(CsvParserMixin, GenericImportParser):
     pass
+
```
Comment (wmorrell) [inline server/edd/load/parsers/generic.py:98]: You have not installed the `pre-commit` hooks. 

1. Run `pre-commit install` from the root of the repo;
2. Then, run `git fetch origin trunk:trunk`;
3. Rebase to `trunk` with `git rebase -i trunk`, and change all the beginning-of-line `pick`s to `edit`;
4. Step through the rebase process following the command line instructions, re-committing each commit with `pre-commit` enabled this time. You'll likely have to fix a few things that `pre-commit` doesn't fix for you.

File: server/edd/load/tests/test_ambr_parser.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+from ..signals import warnings_reported
+from . import factory
+
+
+def test_AmbrExcelParser_success():
+
+    path = ("ambr_export_test_data.xlsx",)
+    parser = parsers.AmbrExcelParser(uuid4())
+
+    with factory.load_test_file(*path) as file:
```
Comment (wmorrell) [inline server/edd/load/tests/test_ambr_parser.py]: Why are you doing argument unpacking here? Just use the path variable, not in a tuple, directly.

File: server/edd/load/tests/test_ambr_parser.py [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+        parsed = parser.parse(file)
+    print(parsed)
```
Comment (wmorrell) [inline server/edd/load/tests/test_ambr_parser.py]: Don't commit tests that use `print`. It's fine to fall back to printing things when trying to debug, but remove it before you commit.

File: server/edd/settings/base.py [hunk 12]
```diff
@@ -273,8 +273,8 @@ def __contains__(self, address):
 # Static files (CSS, JavaScript, Images)
 # https://docs.djangoproject.com/en/dev/howto/static-files/
 # Use the Docker image version hash to distinguish static asset manifest(s)
-with open("/edd.hash") as f:
-    EDD_VERSION_HASH = f.read().strip()
+# with open("/edd.hash") as f:
+EDD_VERSION_HASH = 1287461
```
Comment (wmorrell) [inline server/edd/settings/base.py]: Revert this change. You're hard-coding a specific commit hash used to version the running code. With this change there's no way to easily tell which version of the code you're running anymore.

**Review by wmorrell — CHANGES_REQUESTED** [review]
This needs a lot of work still.

File: Pipfile (hunk 2 above)
Comment (somtirtharoy) [reply Pipfile]: I did this to run my tests locally. I used IntelliJ which was looking for this file in the root directory. It enabled it to create a virtual environment where I could run the tests locally from IntelliJ without having to exec in to a container. I felt that it would be easier to have this in the root directory, but I can remove that.

File: Pipfile.lock (hunk 3 above)
Comment (somtirtharoy) [reply Pipfile.lock]: Same as above. Removing it.

File: server/edd/load/parsers/ambr.py (hunk 4 above)
Comment (somtirtharoy) [reply server/edd/load/parsers/ambr.py:5]: I am using _parse_rows, _parse_row and subsequent validation functions from the GenericParser hence the inheritance. I would like to discuss this with you further. I am converting the multi sheet format to a format that can leverage the GenericImportParser framework without having to reinvent everything.

File: server/edd/load/parsers/core.py (hunk 5 above)
Comment (somtirtharoy) [reply server/edd/load/parsers/core.py]: Yes I will make the changes.

Comment (somtirtharoy) [reply server/edd/load/parsers/core.py]: But I don't still understand how to configure the database and would like some pointers in that direction.

File: server/edd/load/parsers/core.py (hunk 6 above)
Comment (somtirtharoy) [reply server/edd/load/parsers/core.py]: Will make the change

File: server/edd/load/parsers/core.py (hunk 7 above)
Comment (somtirtharoy) [reply server/edd/load/parsers/core.py]: Yes I agree.

File: server/edd/load/parsers/demo.py (hunk 8 above)
Comment (somtirtharoy) [reply server/edd/load/parsers/demo.py]: This file was my prototype file, removing this.

File: server/edd/load/parsers/generic.py (hunk 9 above)
Comment (somtirtharoy) [reply server/edd/load/parsers/generic.py:98]: Okay I will go through this.

File: server/edd/load/tests/test_ambr_parser.py (hunk 10 above)
Comment (somtirtharoy) [reply server/edd/load/tests/test_ambr_parser.py]: Okay.

File: server/edd/load/tests/test_ambr_parser.py (hunk 11 above)
Comment (somtirtharoy) [reply server/edd/load/tests/test_ambr_parser.py]: Sure, will do that.

File: server/edd/settings/base.py (hunk 12 above)
Comment (somtirtharoy) [reply server/edd/settings/base.py]: Yes this was to make the tests work as it was unable to find the edd.hash file locally. I will undo this. Thanks

File: server/edd/load/parsers/core.py (hunk 1 above)
Comment (somtirtharoy) [reply server/edd/load/parsers/core.py:648]: Okay I will think about how to reorder this code so this makes sense or else I will just rename the class.

File: server/edd/load/tests/test_ambr_parser.py [hunk 13]
```diff
@@ -1,21 +1,18 @@
...
 
 
 def test_AmbrExcelParser_success():
 
     path = ("ambr_export_test_data.xlsx",)
+    # path = ("ambr_big_data.xlsx",)
     parser = parsers.AmbrExcelParser(uuid4())
 
-    with factory.load_test_file(*path) as file:
+    with factory.load_test_file(path) as file:
```
Comment (wmorrell) [inline server/edd/load/tests/test_ambr_parser.py:67]: You took off the splat operator, but left `path` at a 1-tuple. Line 9 should be `path = "ambr_export_test_data.xlsx"` (note no parenthesis and comma, so making a string literal and not a 1-tuple).

**Review by wmorrell — CHANGES_REQUESTED** [review]
You've added an excel hidden file `~$ambr_big_data.xlsx`. That should get removed.

**Comment by wmorrell** [comment]
Please update this by rebasing on latest `trunk`, and consolidating all the "wip" commits to what you believe is the final code to merge. It's hard to tell if you've actually resolved many of the comments I have made so far.

File: server/edd/load/parsers/core.py [hunk 14]
```diff
@@ -641,6 +645,127 @@ def _raw_cell_value(self, cell):
...
+                two_cols = sheet[sheet.columns[i : i + 2]]
+
+                # dropping all rows with nan values in the worksheet
+                two_cols = two_cols.dropna()
+                if not two_cols.dropna().empty:
+                    # decimate the data
+                    two_cols = two_cols.iloc[::10, :]
```
Comment (wmorrell) [inline server/edd/load/parsers/core.py]: This does not belong in the parser stage. The parser should read *all* of the data, then pass it along. The code that converts parsed and resolved data into database records should make the determination of which items to drop, and whether to do anything more complicated than a blind, one-in-ten sampling.

File: server/edd/load/parsers/core.py [hunk 15]
```diff
@@ -641,6 +645,127 @@ def _raw_cell_value(self, cell):
...
+    df: pd.DataFrame
+    units: Dict
+    mes_unit_map: Dict
+
+    def __init__(self, name=None, df=None):
+        self.loa_name = name
+        self.df = df
+        # maps naming format of Ambr measurement type to
+        # naming convention in EDD
+        self.mtypes = {
```
Comment (wmorrell) [inline server/edd/load/parsers/core.py]: Don't hard-code these. If the input file uses different names than what is currently in EDD, that translation needs to happen at the data level, and not as hard-coded entries.

File: server/edd/load/parsers/core.py (hunk 6 above)
Comment (wmorrell) [reply server/edd/load/parsers/core.py]: This is still here. Just use `df`, `parsed_df` does not need to exist.

File: server/main/models/core.py [hunk 16]
```diff
@@ -1375,3 +1375,27 @@ def active_in(cls, *, study_id, protocol_id, assay_id=None, id_range=None):
             measurement__assay__protocol_id=protocol_id,
         )
         return active
+
+
+class DefaultUnit(models.Model):
```
Comment (wmorrell) [inline server/main/models/core.py]: This model belongs in the `edd.load` module, not in `main`. It's specific to loading/importing data, not the core data models.

File: server/edd/load/tests/files/defaultunit_data.json [hunk 17]
```diff
@@ -0,0 +1 @@
....defaultunit", "pk": 11, "fields": {"measurement_type": 10, "unit": 18, "protocol": 8, "parser": "Ambr"}}, {"model": "main.defaultunit", "pk": 12, "fields": {"measurement_type": 11, "unit": 18, "protocol": 8, "parser": "Ambr"}}, {"model": "main.defaultunit", "pk": 13, "fields": {"measurement_type": 12, "unit": 1, "protocol": 8, "parser": "Ambr"}}, {"model": "main.defaultunit", "pk": 14, "fields": {"measurement_type": 13, "unit": 17, "protocol": 8, "parser": "Ambr"}}, {"model": "main.defaultunit", "pk": 15, "fields": {"measurement_type": 16, "unit": 17, "protocol": 8, "parser": "Ambr"}}, {"model": "main.defaultunit", "pk": 16, "fields": {"measurement_type": 14, "unit": 17, "protocol": 8, "parser": "Ambr"}}, {"model": "main.defaultunit", "pk": 17, "fields": {"measurement_type": 15, "unit": 17, "protocol": 8, "parser": "Ambr"}}, {"model": "main.defaultunit", "pk": 18, "fields": {"measurement_type": 17, "unit": 17, "protocol": 8, "parser": "Ambr"}}, {"model": "main.defaultunit", "pk": 19, "fields": {"measurement_type": 7, "unit": 16, "protocol": 8, "parser": "Ambr"}}, {"model": "main.defaultunit", "pk": 20, "fields": {"measurement_type": 6, "unit": 15, "protocol": null, "parser": ""}}]
```
Comment (wmorrell) [inline server/edd/load/tests/files/defaultunit_data.json]: Should use a Factory for generating these in the tests that need them. Using a fixture file will hard-code ID values that may not be correct as the code changes.

**Review by wmorrell — CHANGES_REQUESTED** [review]

**Comment by wmorrell** [comment]
Som,

You've added 117 commits to the pull request. You're re-writing a ton of stuff from EDD's history, and reverting quite a few things. Please fix your branch, and force push again.

**Comment by somtirtharoy** [comment]
Hi William, I rebased master on my feature branch and pushed it to remote.
There were conflicts but I picked the changes from the trunk so that I do
not revert anything. I want to confirm that's what you want me to do?

Thanks,

**Comment by wmorrell** [comment]
No, that's precisely the *opposite* of what you should have done. You rebase *your* branch onto trunk, not trunk onto your code.

**Comment by somtirtharoy** [comment]
Oops, okay I will work on it. I didn't want the trunk to be affected.
Thanks. I will have to undo the rebase and then do that again.

Thanks

**Comment by somtirtharoy** [comment]
Hi William, I had to cherry pick my commits and then I rebased it onto the
trunk on my repo and created a new pull request. Please let me know if
everything looks alright and it can be pushed to production.

Thanks,

**Comment by wmorrell** [comment]
Closing this one in favor of the non-draft pull request. Almost there …
